core/src/types/entry.rs

```diff
@@ -18,63 +18,91 @@
+/// use opendal::Metakey;
+/// use opendal::Operator;
+/// # #[tokio::main]
+/// # async fn test(op: Operator) -> Result<()> {
+/// let mut entries = op
+///     .list_with("dir/")
+///     .metakey(Metakey::ContentLength | Metakey::LastModified)
+///     .await?;
+/// for entry in entries {
+///     let meta = entry.metadata();
+///     match meta.mode() {
+///         EntryMode::FILE => {
+///             println!(
+///                 "Handling file {} with size {}",
+///                 entry.path(),
+///                 meta.content_length()
+///             )
+///         }
+///         EntryMode::DIR => {
+///             println!("Handling dir {}", entry.path())
+///         }
+///         EntryMode::Unknown => continue,
+///     }
+/// }
+/// # Ok(())
+/// # }
+/// ```
 #[derive(Clone, Debug)]
 pub struct Entry {
-    /// Path of the entry.
+    /// Path of this entry.
     path: String,
 
-    /// Optional cached metadata
-    metadata: Option<Metadata>,
+    /// Metadata of this entry.
+    metadata: Metadata,
 }
 
 impl Entry {
-    /// Create an entry with .
+    /// Create an entry with metadata.
     ///
     /// # Notes
     ///
-    /// This function is crate internal only. Users don't have public
-    /// methods to construct an entry with arbitrary cached metadata.
-    ///
     /// The only way to get an entry with associated cached metadata
-    /// is `Operator::list` or `Operator::scan`.
-    pub(crate) fn new_with(path: String, metadata: Metadata) -> Self {
-        Self {
-            path,
-            metadata: Some(metadata),
-        }
-    }
-
-    /// Create an [`Entry`] with empty cached metadata.
-    pub fn new(path: &str) -> Self {
-        Self {
-            path: normalize_path(path),
-            metadata: None,
-        }
+    /// is `Operator::list`.
+    pub(crate) fn new(path: String, metadata: Metadata) -> Self {
+        Self { path, metadata }
     }
 
     /// Path of entry. Path is relative to operator's root.
+    ///
     /// Only valid in current operator.
+    ///
+    /// If this entry is a dir, `path` MUST end with `/`
+    /// Otherwise, `path` MUST NOT end with `/`.
     pub fn path(&self) -> &str {
         &self.path
     }
 
     /// Name of entry. Name is the last segment of path.
     ///
-    /// If this entry is a dir, `Name` MUST endswith `/`
-    /// Otherwise, `Name` MUST NOT endswith `/`.
+    /// If this entry is a dir, `name` MUST end with `/`
+    /// Otherwise, `name` MUST NOT end with `/`.
     pub fn name(&self) -> &str {
         get_basename(&self.path)
     }
 
-    /// Get the cached metadata of entry.
-    ///
-    /// # Notes
-    ///
-    /// This function is crate internal only. Because the returning
-    /// metadata could be incomplete. Users must use `Operator::metadata`
-    /// to query the cached metadata instead.
-    pub(crate) fn metadata(&self) -> &Option<Metadata> {
+    /// Fetch metadata of this entry.
+    pub fn metadata(&self) -> &Metadata {
```

Comment:
That's what I want. 😆 So if there is no metakey set, will get a `metadata` with every field being None?

Reply:
If there is no metakey set, metadata will have the values returned by services. `None` here means `NotExist` or `Unknown`. Users can add `metakey` query to make sure that `None` means `NotExist`.

Reply:
I have added some more docs for `Entry::metadata()` call to make it clear.

Reply:
Thanks for adding more docs. May you also update the `list` and `list_with` or `lister` related? Could be finished by just a link to metadata or more.

Reply:
> May you also update the `list` and `list_with` or `lister` related?

Already added.